tests/integration-tests/tests/cloudwatch_logging/test_cloudwatch_logging.py

```diff
@@ -0,0 +1,461 @@
+        compute_statuses = self._run_command_on_computes(status_cmd)
+        hostname_to_status_dict = {hostname: json.loads(status) for hostname, status in compute_statuses.items()}
+        for hostname, status_dict in hostname_to_status_dict.items():
+            self._cluster_log_state["ComputeFleet"][hostname]["agent_status"] = status_dict["status"]
+
+    def _populate_agent_status(self):
+        """Get the cloudwatch agent's status for all the nodes in the cluster."""
+        self._populate_master_agent_status()
+        self._populate_compute_agent_status()
+        LOGGER.debug("After populating agent statuses:\n{}".format(self._dump_cluster_log_state()))
+
+    def _set_cluster_log_state(self):
+        """
+        Get the state of the cluster as it pertains to the CloudWatch logging feature.
+
+        In particular:
+        * Identify which EC2 instances belong to this cluster
+        * Identify which logs are relevant to the MasterServer and ComputeFleet nodes
+        * Identify whether each of a node's relevant logs contain data or not. If they do contain
+          data, save the last line of the file.
+        * Get the CloudWatch agent's status for each node
+        """
+        self._get_initial_cluster_log_state()
+        self._get_relevant_logs()
+        self._populate_log_existence()
+        self._populate_log_emptiness_and_tail()
+        self._populate_agent_status()
+
+
+class Boto3DataRetriever:
+    """Class used to get data from CloudWatch logs."""
+
+    def __init__(self, region):
+        """Save references to the variables the class will need to get log data."""
+        self.region = region
+
+    def get_log_groups(self):
+        """Get list of log groups."""
+        logs_client = boto3.client('logs', region_name=self.region)
+        log_groups = logs_client.describe_log_groups()['logGroups']
+        LOGGER.debug("Log groups: {}\n".format(json.dumps(log_groups, indent=4)))
+        return log_groups
+
+    def get_log_streams(self, log_group_name):
+        """Get list of log streams."""
+        logs_client = boto3.client('logs', region_name=self.region)
+        streams = logs_client.describe_log_streams(logGroupName=log_group_name)['logStreams']
+        LOGGER.debug("Log streams for {group}:\n{streams}".format(
+            group=log_group_name, streams=json.dumps(streams, indent=4)))
+        return streams
+
+    def get_log_events(self, log_group_name, log_stream_name):
+        """Get log events for the given log_stream_name."""
+        logs_client = boto3.client('logs', region_name=self.region)
+        events = logs_client.get_log_events(logGroupName=log_group_name, logStreamName=log_stream_name)
+        LOGGER.debug("Log events for {group}/{stream}:\n{events}".format(
+            group=log_group_name, stream=log_stream_name, events=json.dumps(events, indent=4)))
+        return events
+
+    def get_ec2_instances(self):
```

Comment:
There's probably an easier way to get the information you need.

Reply:
Resolving this because I think it's very closely related to the conversation under `_get_initial_cluster_log_state`.